nearcache.go

```diff
@@ -0,0 +1,692 @@
+	// checkAvailable()
+	// if there is no eviction configured we return if the Near Cache is full and it's a new key.
+	// we have to check the key, otherwise we might lose updates on existing keys.
+	/*
+	   if (evictionDisabled && evictionChecker.isEvictionRequired() && !containsRecordKey(key)) {
+	       return NOT_RESERVED;
+	   }
+	*/
+	rid := rs.nextReservationID()
+	var rec *nearCacheRecord
+	var err error
+	if ups == nearCacheUpdateSemanticWriteUpdate {
+		rec, err = rs.reserveForWriteUpdate(key, keyData, rid)
+	} else {
+		rec, err = rs.reserveForReadUpdate(key, keyData, rid)
+	}
+	if err != nil {
+		return 0, err
+	}
+	if rec == nil || rec.ReservationID() != rid {
+		return nearCacheRecordNotReserved, nil
+	}
+	return rid, nil
+}
+
+func (rs *nearCacheRecordStore) nextReservationID() int64 {
+	return atomic.AddInt64(&rs.reservationID, 1)
+}
+
+func (rs *nearCacheRecordStore) reserveForWriteUpdate(key interface{}, keyData serialization.Data, reservationID int64) (*nearCacheRecord, error) {
+	rs.recordsMu.Lock()
+	defer rs.recordsMu.Unlock()
+	rec, ok := rs.records[key]
+	if !ok {
+		rec, err := rs.createRecord(nil)
+		if err != nil {
+			return nil, err
+		}
+		rec.SetReservationID(reservationID)
+		// initInvalidationMetaData(record, key, keyData);
+	}
+	if rec.reservationID == nearCacheRecordReadPermitted {
+		rec.SetReservationID(reservationID)
+		return rec, nil
+	}
+	// 3. If this record is a previously reserved one, delete it.
+	// Reasoning: CACHE_ON_UPDATE mode has different characteristics
+	// than INVALIDATE mode when updating local near-cache. During
+	// update, if CACHE_ON_UPDATE finds a previously reserved
+	// record, that record is deleted. This is different from
+	// INVALIDATE mode which doesn't delete previously reserved
+	// record and keeps it as is. The reason for this deletion
+	// is: concurrent reservation attempts. If CACHE_ON_UPDATE
+	// doesn't delete previously reserved record, indefinite
+	// read of stale value situation can be seen. Since we
+	// don't apply invalidations which are sent from server
+	// to near-cache if the source UUID of the invalidation
+	// is same with the end's UUID which has near-cache on
+	// it (client or server UUID which has near cache on it).
+	return nil, nil
+}
+
+func (rs *nearCacheRecordStore) reserveForReadUpdate(key interface{}, keyData serialization.Data, reservationID int64) (*nearCacheRecord, error) {
+	key = rs.makeMapKey(key)
+	rs.recordsMu.Lock()
+	defer rs.recordsMu.Unlock()
+	rec, ok := rs.records[key]
+	if ok {
+		return rec, nil
+	}
+	rec, err := rs.createRecord(nil)
+	if err != nil {
+		return nil, err
+	}
+	rec.SetReservationID(reservationID)
+	rs.records[key] = rec
+	// initInvalidationMetaData(record, key, keyData);
+	return rec, nil
+}
+
+func (rs *nearCacheRecordStore) createRecord(value interface{}) (*nearCacheRecord, error) {
+	// assumes recordsMu was locked elsewhere
```

Comment:
Does this method need the mutex to be locked? It does not update rs.records 🤔

Reply:
:thinking: probably not.